src/types_proposal.jl

```diff
@@ -0,0 +1,24 @@
+struct TrackedArray{T,N,A<:AbstractArray{T,N}} <: AbstractArray{T,N}
+    data::A
+    tracker::_Tracker{A}
+    grad::A # todo: grad appears twice
+    TrackedArray{T,N,A}(t::_Tracker{A}, data::A) where {T,N,A} = new(t, data)
+    TrackedArray{T,N,A}(t::_Tracker{A}, data::A, grad::A) where {T,N,A} = new(t, data, grad)
+end
+
+mutable struct _Tracker{T, Pb, Parents}
+    pullback::Pb # a pullback function or nothing for leafs
+    parents::Parents # Tuple of Union{TrackedArray, TrackedReal, TrackedTuple, NotTracked}
+    grad::T # TODO: grad appears twice    
+    ref::UInt32 # currently used by the backpropagation algo
+    isleaf::Bool #TODO: isnt's pullback == nothing enough for the isleaf check?
+    
+    _Tracker{T}(pullback) where T = new(0, pullback, false)
+    _Tracker{T}(pullback, grad::T) where T = new(0, pullback, false, grad)
+    _Tracker{T}(nothing, grad::T) where T = new(0, nothing, true, grad)
+end
+
```

Comment:
Maybe adding more type information for `Pb` and `Parents` would make things more clearer:
```julia
TrackedTypes = Union{TrackedReal, TrackedArray, TrackedTuple, NotTracked}

mutable struct _Tracker{T, Pb, Parents} where {Pb<:Union{Function, Nothing}, Parents<:Tuple{Vararg{TrackedTypes}}}
    ref::UInt32 # currently used by the backpropagation algo, should be the first field for constructors 
    pullback::Pb # pullback function returned by ChainRulesCore.rrule (second returned element)
    parents::Parents # input arguments to to the overloaded operator
    grad::T # gradient (of the same type as the primal data)
    
    _Tracker{T}(pullback, parents) where T = new(0, pullback, parents)
    _Tracker{T}(pullback, parents, grad::T) where T = new(0, pullback, parents, grad)
    # _Tracker{T}(nothing, parents, grad::T) where T = new(0, nothing, parents, grad) # leaf, but not needed anymore, pullback can be nothing above
end

isleaf(tracker::_Tracker) = tracker.pullback === nothing
```

Reply:
I have updated the code accordingly, see bellow. Looks very verbose and I'm not sure if we really get some benefits from making all types explicit.